### ISSUE TITLE
Refactor shortcode parsing

### ISSUE THREAD

**Issue by maxime-rainville** [silverstripe/silverstripe-framework#11260]
Our approach to managing short code is dated and uses many anti-patterns.

Each individual shortcode class must implement the [`ShortcodeHandler`](https://github.com/silverstripe/silverstripe-framework/blob/5/src/View/Parsers/ShortcodeHandler.php) interface. This interface requires the usage of `static` method. Nowadays, we prefer to use the singleton pattern and use instance methods.

Short code must be registered in `_config.php` by making this kind of calls:
https://github.com/silverstripe/silverstripe-framework/blob/8886a3a93ddf5d9ad0b677f7403ca600894bbe00/_config.php#L14-L15

Registering shortcodes should be handle via the Injector.

Most of the logic in [`ShortcodeParser`](https://github.com/silverstripe/silverstripe-framework/blob/5/src/View/Parsers/ShortcodeParser.php) is dedicated to parsing BBCode shortcodes. There's libraries around for this exact purpose:
- [thunderer/shortcode](https://packagist.org/packages/thunderer/shortcode)
- [s9e/text-formatter](https://packagist.org/packages/s9e/text-formatter)

We should consider adopting them instead of maintaining our own parser.

**Comment by maxime-rainville** [comment]
I had a very quick look at the two libraries:
- [thunderer/shortcode](https://packagist.org/packages/thunderer/shortcode) seems very aligned with our current approach, but it hasn't been updated in two years.
- [s9e/text-formatter](https://packagist.org/packages/s9e/text-formatter) has been updated in the last month but it seems like quite a departure from the current approach. It seems to prefer a "template" approach that might not be suitable for our uses case.

**Comment by GuySartorelli** [comment]
> Each individual shortcode class must implement the [ShortcodeHandler](https://github.com/silverstripe/silverstripe-framework/blob/5/src/View/Parsers/ShortcodeHandler.php) interface.

That's not true - shortcode parsers don't even need to be classes. You only register a callback. It could even be an anonymous function with the current setup.

But yes, I agree we should offload this to a third-party library if one is available and appropriate.

Alternatively we should allow registering them with yaml config, even if we change nothing else about our current approach.
